### ISSUE TITLE
Differences in solutions for the same optimize function

### ISSUE THREAD

**Issue by fpauperio** [opencobra/cobrapy#1025]
### Question
Hello everyone,

   Recently I was trying to use cobrapy to use a E.coli model and I did a loop that would knockout some combination of genes. I used the examples from the documentation for that. However, I noticed that the objective value is different when I use the "with model as model:" or if I create the model variable from the file each time I knock out the combination. I want to understand why the results are different, if they have the same medium, the same genes knocked out and the same objective function.

Example: 
model = cobra model from file 
with model as model:
     gene knockout 1
     model.optimize
     solution.objective_value

with model as model:
     gene knockout 2
     model.optimize
     solution.objective_value

#is not the same as

model = cobra model from file 
with model as model:
     gene knockout 1
     model.optimize
     solution.objective_value

model = cobra model from file 
with model as model:
     gene knockout 2
     model.optimize
     solution.objective_value

Thank you very much,
Francisco

**Comment by Midnighter** [comment]
Hi @fpauperio,

It would help if you format the relevant sections as code since the indentation of code blocks is now unclear. You can do so either through the interface or using triple backticks

    ```

Also, you don't show how the solution object is defined. Since you use an _E. coli_ model that is probably publicly available it would help to mention which model exactly and which genes you are using in your example.

**Comment by fpauperio** [comment]
Hello again, 

 and sorry for the lack of information given previously! I am using the e.coli model from this paper: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3261703/

 The objective function is the default of the model, which is called " biomass", which calculates the amount of biomass produced.
Maximize
1.0*Ec_biomass_iJO1366_core_53p95M - 1.0*Ec_biomass_iJO1366_core_53p95M_reverse_e94eb
```
model = cobra.io.read_sbml_model("e.coli_palsson_2011.xml")
with model as model:
    model.genes.b0929.knock_out()
    solution =model.optimize()
    print(solution.objective_value)

with model as model:
    model.genes.b0928.knock_out()
    solution =model.optimize()
    print(solution.objective_value)
    a = solution.objective_value

#is not the same as

model = cobra.io.read_sbml_model("e.coli_palsson_2011.xml")
with model as model:
    model.genes.b0929.knock_out()
    solution =model.optimize()
    print(solution.objective_value)

model = cobra.io.read_sbml_model("e.coli_palsson_2011.xml")
with model as model:
    model.genes.b0928.knock_out()
    solution =model.optimize()
    print(solution.objective_value)
    b = solution.objective_value
```

In summary, there is a difference in the objective value of a second knockout using "with" when compared to reload the model each time. A different than B, although only the b0928 gene is knocked out.

Thank you,
happy holidays!

**Comment by cdiener** [comment]
Unfortunately I can't reproduce this. Here is what I get:

```bash
In [6]: model = cobra.io.read_sbml_model("/home/cdiener/Downloads/msb201165-s3.xml") 
   ...: model.solver = "glpk" 
   ...: with model as model: 
   ...:     model.genes.b0929.knock_out() 
   ...:     solution =model.optimize() 
   ...:     print(solution.objective_value) 
   ...:  
   ...: with model as model: 
   ...:     model.genes.b0928.knock_out() 
   ...:     solution =model.optimize() 
   ...:     print(solution.objective_value) 
   ...:     a = solution.objective_value 
   ...:  
   ...: #is not the same as 
   ...:  
   ...: model = cobra.io.read_sbml_model("/home/cdiener/Downloads/msb201165-s3.xml") 
   ...: model.solver = "glpk" 
   ...: with model as model: 
   ...:     model.genes.b0929.knock_out() 
   ...:     solution =model.optimize() 
   ...:     print(solution.objective_value) 
   ...:  
   ...: model = cobra.io.read_sbml_model("/home/cdiener/Downloads/msb201165-s3.xml") 
   ...: model.solver = "glpk" 
   ...: with model as model: 
   ...:     model.genes.b0928.knock_out() 
   ...:     solution =model.optimize() 
   ...:     print(solution.objective_value) 
   ...:     b = solution.objective_value 
   ...:                                                                                                                                                                       
0.9823718127269798
3.698865253144943e-15
0.9823718127269798
-1.4511215425957117e-16
```
Are you concerned about the very small values being different? For all floating point solvers all flux values will only be accurate by about the value indicated by `model.tolerance`. This is in the nature of floating point operations. It is generally not possibly to clearly say whether a flux is exactly zero (https://bitbashing.io/comparing-floats.html).
```

**Comment by cdiener** [comment]
Feel free to reopen if you think the issue is not resolved. Thanks!
